### ISSUE TITLE
child_process.zig, can't spawn executable from env with a cwd

### ISSUE THREAD

**Issue by GreyDodger** [ziglang/zig#3701]
I'd like to spawn a process cl.exe, which can be found in a directory in my environment vars, and I'd like the cwd of that new process to be in a specific directory. This code in child_process.zig (line 567) seems to make this unnecessarily impossible. 

```
        // the cwd set in childprocess is in effect when choosing the executable path
        // to match posix semantics
        const app_name = x: {
            if (self.cwd) |cwd| {
                const resolved = try fs.path.resolve(self.allocator, [_][]const u8{ cwd, self.argv[0] });
                defer self.allocator.free(resolved);
                break :x try cstr.addNullByte(self.allocator, resolved);
            } else {
                break :x try cstr.addNullByte(self.allocator, self.argv[0]);
            }
        };
        defer self.allocator.free(app_name);

```

This will combine my executable with the cwd and then error when that executable does not exist. I could just pass in the absolute path of cl.exe but that's the microsoft visual c++ compiler, and that thing is unreasonably hard to find.

**Comment by daurnimator** [comment]
> which can be found in a directory in my environment vars,

Do your environment vars contain a relative directory? Otherwise I don't see how this would be an issue.

**Comment by GreyDodger** [comment]
No, my environment vars does not contain relative directories, the issue is that I wan't cl.exe to spawn inside a certain directory (because it spits out a bunch of files), but when I specify that new cwd, zig thinks cl.exe lives in that cwd instead of somewhere in my environment vars.

**Comment by marler8997** [comment]
Might be fixed by https://github.com/ziglang/zig/pull/2770

**Comment by GreyDodger** [comment]
Does not fix. Just to clarify.
```
// this works
const std = @import("std");
const ChildProcess = std.ChildProcess;
const direct_allocator = std.heap.direct_allocator;
pub fn main() !void {
    const child_process = try ChildProcess.init([_][]const u8{ "cl" }, direct_allocator);
    var term = try child_process.spawnAndWait();
}
```
```
// this does not work
//error: InvalidName
//C:\Users\busin\dropbox\prg\programs\zig\lib\zig\std\os\windows.zig:783:37: 0x7ff6681ed655 in //std.os.windows.CreateProcessW (executable.obj)
//            ERROR.FILE_NOT_FOUND => return error.FileNotFound,

const std = @import("std");
const ChildProcess = std.ChildProcess;
const direct_allocator = std.heap.direct_allocator;
pub fn main() !void {
    const child_process = try ChildProcess.init([_][]const u8{ "cl" }, direct_allocator);
    child_process.cwd = "C:/prg/zigtest";
    var term = try child_process.spawnAndWait();
}
```
calling child_process.spawnAndWait() on windows eventually calls `windows.CreateProcessW(app_name, ...)`, with the app_name being the cwd and the first program arg concatenated, which according to windows documentation https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw so it doesn't bother searching my path. I want the eventual win32 call to be `CreateProcessW(app_name = "cl", ...., lpCurrentDirectory = "C:/prg/zigtest")`

The comment in the code above
```        
        // the cwd set in childprocess is in effect when choosing the executable path
        // to match posix semantics
```
makes me think that although it's an arbitrary limitation for windows it's still intentional, I mainly opened this issue to confirm this, or to see if there isn't some obvious work around for this problem. (besides manually passing the absolute path to cl.exe")

**Comment by stratact** [comment]
I don't think it's a limitation of Windows per se, but more so that Windows NT systems naturally call executable from the current directory first when using the command line from what I recall. Zig probably has to account for the executable in cwd first for Windows environments while acting different for POSIX. That might make the API a bit inconsistent though...

**Comment by matu3ba** [comment]
One workaround, which also allows more control by the user is to offer 1. removing and 2. filtering of environment variables
as suggest by https://wiki.sei.cmu.edu/confluence/display/c/ENV03-C.+Sanitize+the+environment+when+invoking+external+programs

This could either work via optional function pointer or optional comptime-provided function.
However, offering more selectable defaults than "filter all" or "filter nothing" would be too tedious on maintenance and limit user choices.

**Comment by nexusprime-here** [comment]
the same here
